### DISCUSSION TITLE
[doc] PMD Rule documentation throws 404 for v5.4.1

### DISCUSSION THREAD

**Discussion by ravikiranj** [pmd/pmd#3041] [Q&A] [answered]
**Affects PMD Version:** 5.4.1

Hi there, I'm currently experimenting with an older PMD jar whose version is v5.4.1 and I noticed that all the links in the HTML report threw 404s. For example, https://pmd.github.io/pmd-5.4.1/pmd-java/rules/java/design.html#UseVarargs.

I was able to refer to the same rule by browsing v5.4.6 doc (https://pmd.github.io/pmd-5.4.6/pmd-java/rules/java/design.html#UseVarargs). 

- Is there a reason why certain versions have the documentation missing on the github pages? 
- If we were to upgrade to the latest version, can we rely on the rule links to work in the future?

**Comment by adangel** [comment]
Hi,
yes, the old versions have been removed - basically for a very simple reason: webspace (and github pages deployment time and space limit). You can be sure, that the latest version (current version 6.2.0) is available. To view the old documentation page, you can download the docs archive from https://sourceforge.net/projects/pmd/files/pmd/ . You can see in the source repo of [pmd.github.io](https://github.com/pmd/pmd.github.io/) which versions are available online - currently it is always at least the last version (the micro/bugfix release) of each minor version.

May I ask you, why you are specifically trying out this old version? Is there a feature, that we accidentally removed in a later version?

Regards,
Andreas

**Comment by ravikiranj** [comment]
Thank you for the quick response. I'm currently investigating the effort involved in upgrading the library and checking if we can use our current version as is and found that reports were broken.

Let's say a few versions down the line, even 6.x.x becomes obsolete and the links are broken due to constraints you mentioned. Do you think it makes sense to not link them at all in the first place similar to checkstyle reports? Alternatively, if the docs can be locally hosted, an option to override the base url within the ant task would also be a great idea.

**Comment by jsotuyod** [comment]
> I'm currently investigating the effort involved in upgrading the library and checking if we can use our current version as is and found that reports were broken.

That's interesting... what errors are you hitting? 5.8.1 (last 5.x release) should be completely compatible.

Even 6.x releases should work without problems except for the rule `UseSingleton` being completely removed (it has been deprecated for a long time, the rule `UseUtilityClass` is a drop-in replacement). You may see deprecation warnings when using other old rules / rulesets, but it will keep on working all the way up to 7.0.0 giving you plenty of time to upgrade the config.

If you are seeing other breaking changes, please provide full detail. We are most interested on it.

**Comment by ravikiranj** [comment]
Hi Juan,
>I'm currently investigating the effort involved in upgrading the library and checking if we can use our current version as is and found that reports were broken.

>>That's interesting... what errors are you hitting? 5.8.1 (last 5.x release) should be completely compatible.

Sorry for the misunderstanding! I meant to write that the report links w.r.t rules were broken on 5.4.1 (https://pmd.github.io/pmd-5.4.1/* 404s).

For the short term, I'm trying to fix the broken report links with the current version, i.e., 5.4.1 (locally hosted docs or auto redirector chrome/FF plugin). As a long term fix, we will upgrade to the latest and greatest PMD and port our config/rules to match the new one's.

**Comment by adangel** [comment, marked as answer]
I've now made the old documentation available at http://pmd.sourceforge.net/archive.html

The 404 page of pmd.github.io will now redirect automatically (if javascript is enabled) to the documentation in the archive.

So, the link https://pmd.github.io/pmd-5.4.1/pmd-java/rules/java/design.html#UseVarargs works again.
